modules/apps/headless/headless-portal-instances/headless-portal-instances-impl/src/main/java/com/liferay/headless/portal/instances/internal/resource/v1_0/PortalInstanceResourceImpl.java

```diff
@@ -27,4 +41,155 @@
+
+		List<PortalInstance> portalInstances = new ArrayList<>();
+
+		for (Company company : companies) {
+			if (skipDefaultCompany &&
+				(_portalInstancesLocalService.getDefaultCompanyId() ==
+					company.getCompanyId())) {
+
+				continue;
+			}
+
+			portalInstances.add(_getPortalInstance(company));
+		}
+
+		return Page.of(portalInstances);
+	}
+
+	@Override
+	public PortalInstance patchPortalInstance(
+			String webId, PortalInstance portalInstance)
+		throws Exception {
+
+		Company company = _companyLocalService.getCompanyByWebId(webId);
+
+		String domain = portalInstance.getDomain();
+		String virtualHostname = portalInstance.getVirtualhost();
+
+		if (((domain == null) || domain.equals(company.getMx())) &&
+			((virtualHostname == null) ||
+			 virtualHostname.equals(company.getVirtualHostname()))) {
+
+			throw new ValidationException(
+				"Provide a valid new value for one of the modifiable fields, " +
+					"domain or virtual host");
+		}
+
+		if (domain == null) {
+			domain = company.getMx();
+		}
+
+		if (virtualHostname == null) {
+			virtualHostname = company.getVirtualHostname();
+		}
+
+		return _getPortalInstance(
+			_companyLocalService.updateCompany(
+				company.getCompanyId(), virtualHostname, domain,
+				company.getMaxUsers(), company.isActive()));
+	}
+
+	@Override
+	public PortalInstance postPortalInstance(PortalInstance portalInstance)
+		throws Exception {
+
+		Company company = _companyLocalService.addCompany(
+			portalInstance.getWebId(), portalInstance.getVirtualhost(),
+			portalInstance.getDomain(), false, 0, true);
+
+		_portalInstancesLocalService.initializePortalInstance(
+			ServletContextPool.get(StringPool.BLANK), company.getWebId());
+
+		_portalInstancesLocalService.synchronizePortalInstances();
+
+		return _getPortalInstance(company);
+	}
+
+	@Override
+	public void putPortalInstanceActivate(String webId) throws Exception {
+		Company company = _companyLocalService.getCompanyByWebId(webId);
+
+		if (company.isActive()) {
+			throw new ValidationException("Portal instance already active");
+		}
+
+		_companyLocalService.updateCompany(
+			company.getCompanyId(), company.getVirtualHostname(),
+			company.getMx(), company.getMaxUsers(), true);
+	}
+
+	@Override
+	public void putPortalInstanceDeactivate(String webId) throws Exception {
+		Company company = _companyLocalService.getCompanyByWebId(webId);
+
+		if (_portalInstancesLocalService.getDefaultCompanyId() ==
+				company.getCompanyId()) {
+
+			throw new ValidationException(
+				"Select other default portal instance before deactivating it");
+		}
+
+		if (!company.isActive()) {
+			throw new ValidationException(
+				"Portal instance already deactivated");
+		}
+
+		_companyLocalService.updateCompany(
+			company.getCompanyId(), company.getVirtualHostname(),
+			company.getMx(), company.getMaxUsers(), false);
+	}
+
+	private PortalInstance _getPortalInstance(Company company) {
+		PortalInstance portalInstance = new PortalInstance();
```

Comment:
Good try, Brian likes this pattern:

return new PortalInstance() {
{
active = company.isActive;
...
}
};

Reply:
Ok Will do it